### PR TITLE
FolderWatcher (linux): optimize slotAddFolderRecursive

### PR DESCRIPTION
**Pull request by ogoffart**
- Instead of putting all the pathes in a big QStringList and then process them,
  do the processing as we iterate with QDirIterator
- Since we know that the parent directory is not ignored, we can simplify the
  ignore check by not checking the part before (i.e: we call
  ExcludedFiles::isExcluded with the parent 'path' as base path) This is what
  speeds up considerably the function.

Note that the OWNCLOUD_TEST check is already there in FolderWatcher::pathIsIgnored
so the behavior is not changed

### CODE REVIEW

**Comment by ckamm** [comment]
@ogoffart I haven't read this in detail yet, but wouldn't the change in exclude check break patterns that include the path ("foo/bar/file" wouldn't match if we feed only "file" inside "foo/bar/")? I'm not sure, in general, if we aim to support patterns like that.

**Comment by ogoffart** [comment]
@ckamm  good point

**Comment by guruz** [comment]
I'm actually thinking if we should stop supporting this kind of exclusion pattern?

1) My https://github.com/owncloud/client/pull/4999 does not support it either (yet)
2) There is a bug about this that it is broken anyway: https://github.com/owncloud/client/issues/4628

**Comment by ckamm** [comment]
@ogoffart @guruz But wouldn't this change still break full-absolute paths, like ignoring "/A/B/C"? Because "/A/B/" is not excluded, and "C" doesn't match any pattern on its own.

File: test/testinotifywatcher.cpp [hunk 1]
```diff
@@ -11,18 +11,23 @@
...
+{
+    Q_OBJECT
 private slots:
-    void initTestCase() {
-        qsrand(QTime::currentTime().msec());
+    // Test the recursive path listing function lists everything
+    void testAddFolderRecursiveHelper() {
+        QTemporaryDir tmpDir;
 
-        _root = QDir::tempPath() + "/" + "test_" + QString::number(qrand());
```
Comment (ckamm) [inline test/testinotifywatcher.cpp]: Why did you remove the test_1231 subdirectory?

File: src/gui/folderwatcher_linux.cpp [hunk 2]
```diff
@@ -88,41 +64,47 @@ void FolderWatcherPrivate::inotifyRegisterPath(const QString& path)
...
+
+    QDir::Filters filter = QDir::Dirs | QDir::NoDotAndDotDot | QDir::NoSymLinks | QDir::Hidden;
+    const QStringList pathes = dir.entryList(filter);
+    for (auto constIterator = pathes.constBegin(); constIterator != pathes.constEnd(); ++constIterator) {
+        const QString subfolder = path + QLatin1String("/") + (*constIterator);
+        QDir folder(subfolder);
+        if (folder.exists() && !watchedFolders.contains(subfolder)) {
+#ifndef OWNCLOUD_TEST // InotifyWatcherTest is not interested in ignored files and does not link against the folder
+            if( _parent->_folder->syncEngine().excludedFiles().isExcluded(
+                    subfolder, path, _parent->_folder->ignoreHiddenFiles())) {
```
Comment (ckamm) [inline src/gui/folderwatcher_linux.cpp]: I think for correctness this must always pass in the full folder-relative path. (see other comment)

**Comment by guruz** [comment]
@ckamm I've never seen full absolute paths in the exclude. What is it absolute to? The system root or the sync folder root?

**Comment by ckamm** [comment]
@guruz I meant absolute to the sync folder root. For example the "/Music/iTunes/Album Artwork" pattern that @dragotin recommended in #4628.

File: test/testinotifywatcher.cpp (hunk 1 above)
Comment (ogoffart) [inline test/testinotifywatcher.cpp]: I'm using QTemporaryDir which take care of that instead of manually creating and destroying the directory.

**Comment by ckamm** [comment]
I've created #5275 for discussion about what the desired behavior for exclude patterns even is.

**Comment by guruz** [comment]
I wrote my personal conclusion as comment on https://github.com/owncloud/client/issues/5275#issuecomment-262519711
How would we integrate this with this PR here? (Because we want to fix https://github.com/owncloud/client/issues/4628 too)

**Comment by ckamm** [comment]
@ogoffart I'd close this?
